### PR TITLE
chore: upgrade for opentelemtry js v0.12.0

### LINKED ISSUE #188: Upgrade to opentelemetry 0.12.0
A new verdion of opentelemetry was released, but because it is a 0.x version it won't get updated unless it is allowed in the dependency file.

See https://github.com/open-telemetry/opentelemetry-js/releases/tag/v0.12.0 for the breaking changes 

I asked about it before, but it would be great to pin the version number of the provided packages to the opentelemetry version, to ease the matching of the packages.

### PR DESCRIPTION
**Pull request by aabmass**
- [x] Make sure to open an issue as a [bug/issue/feature](https://github.com/GoogleCloudPlatform/opentelemetry-operations-js/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Part of #188 

Upgrade OpenTelemetry peer dependencies to v0.12.0.

- `Distribution` was removed altogether
- Some other fixes for slight API changes
- I think the codecov drop is flaky since I deleted code?

### CODE REVIEW

File: packages/opentelemetry-cloud-trace-propagator/test/CloudPropagator.test.ts [hunk 1]
```diff
@@ -75,9 +73,9 @@ describe('CloudPropagator', () => {
     it('should extract context of a sampled span from carrier', () => {
       carrier[X_CLOUD_TRACE_HEADER] =
         'd4cda95b652f4a1592b449d5929fda1b/7929822056569588882;o=1';
-      const extractedSpanContext = getExtractedSpanContext(
```
Comment (aabmass) [inline packages/opentelemetry-cloud-trace-propagator/test/CloudPropagator.test.ts:78]: This function was deleted altogether in favor of using the active span

File: packages/opentelemetry-cloud-monitoring-exporter/src/monitoring.ts [hunk 2]
```diff
@@ -140,7 +140,7 @@ export class MetricExporter implements IMetricExporter {
     cb(ExportResult.SUCCESS);
   }
 
-  shutdown(): void {}
+  async shutdown(): Promise<void> {}
```
Comment (dashpole) [inline packages/opentelemetry-cloud-monitoring-exporter/src/monitoring.ts:143]: just curious, where did this change come from?

**Review by dashpole — APPROVED** [review]

File: packages/opentelemetry-cloud-trace-exporter/test/transform.test.ts [hunk 3]
```diff
@@ -130,6 +130,7 @@ describe('transform', () => {
   });
 
   it('should drop unknown attribute types', () => {
+    // @ts-expect-error
```
Comment (mayurkale22) [inline packages/opentelemetry-cloud-trace-exporter/test/transform.test.ts:133]: Why do we need this?

File: packages/opentelemetry-cloud-trace-exporter/src/trace.ts [hunk 4]
```diff
@@ -80,7 +80,7 @@ export class TraceExporter implements SpanExporter {
     resultCallback(result);
   }
 
-  shutdown(): void {}
+  async shutdown(): Promise<void> {}
```
Comment (mayurkale22) [inline packages/opentelemetry-cloud-trace-exporter/src/trace.ts:83]: I think we should add some logic here to set the shutdown flag to true when this method is called and in export return immediately when the flag is set to true. Currently, shutdown operation looks more like a no-op. WDYT? Perhaps, we can do this in separate PR.

File: packages/opentelemetry-cloud-monitoring-exporter/src/monitoring.ts (hunk 2 above)
Comment (aabmass) [reply packages/opentelemetry-cloud-monitoring-exporter/src/monitoring.ts:143]: They changed the interface in the SDK: https://github.com/open-telemetry/opentelemetry-js/pull/1439

File: packages/opentelemetry-cloud-trace-exporter/src/trace.ts (hunk 4 above)
Comment (aabmass) [reply packages/opentelemetry-cloud-trace-exporter/src/trace.ts:83]: Makes sense, I haven't really look into it yet -- tracking in https://github.com/GoogleCloudPlatform/opentelemetry-operations-js/issues/190

File: packages/opentelemetry-cloud-trace-exporter/test/transform.test.ts (hunk 3 above)
Comment (aabmass) [reply packages/opentelemetry-cloud-trace-exporter/test/transform.test.ts:133]: The test is asserting that unsupported attribute types (like object here) are dropped, but it creates a TS error since it is not allowed. In this case, we expect the TS error but want to validate the behavior too
